chia-keys/src/secret_key.rs

```diff
@@ -0,0 +1,302 @@
+use crate::derivable_key::DerivableKey;
+use crate::public_key::PublicKey;
+use bls12_381_plus::{G1Projective, Scalar};
+use hkdf::HkdfExtract;
+use num_bigint::BigUint;
+use sha2::{Digest, Sha256};
+
+#[derive(PartialEq, Eq, Debug)]
+pub struct SecretKey(pub(crate) Scalar);
+
+fn flip_bits(input: [u8; 32]) -> [u8; 32] {
+    let mut ret = [0; 32];
+    for i in 0..32 {
+        ret[i] = input[i] ^ 0xff;
+    }
+    ret
+}
+
+fn ikm_to_lamport_sk(ikm: &[u8; 32], salt: &[u8; 4]) -> [u8; 255 * 32] {
+    let mut extracter = HkdfExtract::<Sha256>::new(Some(salt));
+    extracter.input_ikm(ikm);
+    let (_, h) = extracter.finalize();
+
+    let mut output = [0_u8; 255 * 32];
+    h.expand(&[], &mut output).unwrap();
+    output
+}
+
+fn sha256(bytes: &[u8]) -> [u8; 32] {
+    let mut hasher = Sha256::new();
+    hasher.update(bytes);
+    hasher.finalize().try_into().unwrap()
+}
+
+impl SecretKey {
+    pub fn from_seed(seed: &[u8]) -> SecretKey {
+        // described here:
+        // https://eips.ethereum.org/EIPS/eip-2333#derive_master_sk
+        assert!(seed.len() >= 32);
+
+        const SALT: &[u8] = b"BLS-SIG-KEYGEN-SALT-";
+        let mut extracter = HkdfExtract::<sha2::Sha256>::new(Some(SALT));
+        extracter.input_ikm(seed);
+        extracter.input_ikm(&[0_u8]);
+        let h = extracter.finalize().1;
+
+        // TODO: = uninitialized
+        let mut sk = [0_u8; 48];
+        h.expand(&[0_u8, 48], &mut sk)
+            .expect("failed to generate secret key");
+        SecretKey(Scalar::from_okm(&sk))
+    }
+
+    pub fn from_bytes(b: &[u8; 32]) -> Option<SecretKey> {
+        let t = [
+            b[31], b[30], b[29], b[28], b[27], b[26], b[25], b[24], b[23], b[22], b[21], b[20],
+            b[19], b[18], b[17], b[16], b[15], b[14], b[13], b[12], b[11], b[10], b[9], b[8], b[7],
+            b[6], b[5], b[4], b[3], b[2], b[1], b[0],
+        ];
+        Scalar::from_bytes(&t).map(SecretKey).into()
+    }
+
+    pub fn to_bytes(&self) -> [u8; 32] {
+        let mut bytes = self.0.to_bytes();
+        bytes.reverse();
+        bytes
+    }
+
+    pub fn public_key(&self) -> PublicKey {
+        PublicKey(G1Projective::generator() * self.0)
+    }
+
+    fn to_lamport_pk(&self, idx: u32) -> [u8; 32] {
```

Comment:
Okay, this stuff I am not familiar with.

Reply:
this is more or less a verbatim port from `bls-signatures`